File: ulyp-ui/src/main/java/com/ulyp/ui/renderers/RenderedObject.java
package com.ulyp.ui.renderers;

import com.ulyp.core.Type;
import com.ulyp.core.printers.*;
import com.ulyp.ui.RenderSettings;
import javafx.scene.text.TextFlow;

public abstract class RenderedObject extends TextFlow {

    private final Type type;

    protected RenderedObject(Type type) {
        this.type = type;
    }

    public static RenderedObject of(ObjectRepresentation repr, RenderSettings renderSettings) {
        RenderedObject objectValue;

        // TODO replace with map
        if (repr instanceof StringObjectRepresentation) {

            objectValue = new RenderedStringObject((StringObjectRepresentation) repr, repr.getType(), renderSettings);
        } else if (repr instanceof NullObjectRepresentation) {

            objectValue = new RenderedNull(renderSettings);
        } else if (repr instanceof NotRecordedObjectRepresentation) {

            objectValue = new RenderedNotRecordedObject(renderSettings);
        } else if (repr instanceof NumberObjectRepresentation) {

            objectValue = new RenderedNumber((NumberObjectRepresentation) repr, repr.getType(), renderSettings);
        } else if (repr instanceof ObjectArrayRepresentation) {

            objectValue = new RenderedObjectArray((ObjectArrayRepresentation) repr, renderSettings);
        } else if (repr instanceof CollectionRepresentation) {

            objectValue = new RenderedCollection((CollectionRepresentation) repr, renderSettings);
        } else if (repr instanceof MapEntryRepresentation) {

            objectValue = new RenderedMapEntry((MapEntryRepresentation) repr, renderSettings);
        } else if (repr instanceof ClassObjectRepresentation) {

            objectValue = new RenderedClassObject((ClassObjectRepresentation) repr, renderSettings);
        } else if (repr instanceof MapRepresentation) {

            objectValue = new RenderedMap((MapRepresentation) repr, renderSettings);
        } else if (repr instanceof IdentityObjectRepresentation) {

            objectValue = new RenderedIdentityObject((IdentityObjectRepresentation) repr, renderSettings);
        } else if (repr instanceof ThrowableRepresentation) {

            objectValue = new RenderedThrowable((ThrowableRepresentation) repr, renderSettings);
        } else if (repr instanceof EnumRepresentation) {

            objectValue = new RenderedEnum((EnumRepresentation) repr, renderSettings);
        } else if (repr instanceof ToStringPrintedRepresentation) {

            objectValue = new RenderedToStringPrinted((ToStringPrintedRepresentation) repr, renderSettings);
        } else if (repr instanceof DateRepresentation) {

            objectValue = new RenderDate((DateRepresentation) repr, renderSettings);
        } else if (repr instanceof BooleanRepresentation) {

            objectValue = new RenderedBoolean((BooleanRepresentation) repr, renderSettings);
        } else {

            throw new RuntimeException("Not supported for rendering: " + repr);
        }

        objectValue.getChildren().forEach(node -> {
            node.getStyleClass().add("ulyp-ctt");
            node.getStyleClass().add("ulyp-ctt-object-repr");
        });
        return objectValue;
    }
}

File: ulyp-ui/src/main/kotlin/com/ulyp/ui/renderers/RenderedObject.kt
package com.ulyp.ui.renderers

import com.ulyp.core.Type
import com.ulyp.core.recorders.*
import com.ulyp.ui.RenderSettings
import javafx.scene.Node
import javafx.scene.text.TextFlow
import java.util.function.Consumer

abstract class RenderedObject protected constructor(private val type: Type?) : TextFlow() {
    companion object {
        @JvmStatic
        fun of(record: ObjectRecord, renderSettings: RenderSettings?): RenderedObject {
            val objectValue = when (record) {
                is StringObjectRecord -> RenderedString(record, record.getType(), renderSettings)
                is CharObjectRecord -> RenderedChar(record, record.getType(), renderSettings!!)
                is NullObjectRecord -> RenderedNull(renderSettings)
                is NotRecordedObjectRecord -> RenderedNotRecordedObject(renderSettings!!)
                is NumberRecord -> RenderedNumber(record.numberPrintedText, record.getType(), renderSettings!!)
                is FileRecord -> RenderedFile(record.path, record.getType(), renderSettings!!)
                is ObjectArrayRecord -> RenderedObjectArray(record, renderSettings)
                is CollectionRecord -> RenderedCollection(record, renderSettings!!)
                is MapEntryRecord -> RenderedMapEntry(record, renderSettings)
                is ClassObjectRecord -> RenderedClassObject(record, renderSettings!!)
                is MapRecord -> RenderedMap(record, renderSettings!!)
                is IdentityObjectRecord -> RenderedIdentityObject(record, renderSettings!!)
                is ThrowableRecord -> RenderedThrowable(record, renderSettings!!)
                is EnumRecord -> RenderedEnum(record, renderSettings!!)
                is PrintedObjectRecord -> RenderedPrintedObject(record, renderSettings!!)
                is DateRecord -> RenderDate(record, renderSettings!!)
                is BooleanRecord -> RenderedBoolean(record, renderSettings!!)
                is OptionalRecord -> RenderedOptional(record, renderSettings!!)
                else -> throw RuntimeException("Not supported for rendering: $record")
            }
            objectValue.children.forEach(Consumer { node: Node ->
                node.styleClass.add("ulyp-ctt")
            })
            return objectValue
        }
    }
}